_posts/2023-04-30-writing-code-with-llm-dont-forget-kani.md

```diff
@@ -0,0 +1,306 @@
+---
+layout: post
+title: "Writing Code with LLM? Don't Forget Kani"
+---
+
+Kani is a verification tool that can help you prove properties about your Rust code.
+To learn more about Kani, check out [the Kani tutorial](https://model-checking.github.io/kani/kani-tutorial.html) and our [previous blog posts](https://model-checking.github.io/kani-verifier-blog/).
+
+
+Recently, ChatGPT---OpenAI’s AI chatbot built on top of large language models(LLM)---has taken the world by storm.
+People use ChatGPT to write blog posts, do homework, and even write code!
+That’s right---LLMs like ChatGPT can be used to write code, since code is essentially just another kind of text.
+And with ChatGPT, all you need to do is provide a text prompt describing the program you want, and it will generate the code for you - along with an explanation of how it arrived at the solution.
+
+We were also eager to explore the possibilities of using LLMs for daily coding tasks, so we decided to put ChatGPT to the test.
+In this blog post, we'll share our experiences with using ChatGPT and Kani to solve example coding problems, and explore how these technology could revolutionize the way we approach software development.
+The experiment we did include the following three tasks.
+
+1. We write prompt and generate (unverified!) code with ChatGPT.
+2. We verify the generated code to see if they are correct with Kani.
+3. We use both ChatGPT and Kani to generate verified code.
+
+Our result showed that, ChatGPT is easily to use and can efficient generate code from text description. 
+While the generated code is not always sound, program verifiers like Kani can effectively detect bugs in the generated code and help ChatGPT fix them.
+
+## Code Generation using ChatGPT
+We demonstrate the three tasks with two examples.
+Let's start our first task: can we write code using prompt with ChatGPT?
+
+#### Example 1.a
+The first example is a toy problem of writing a rust function to compute the average of two integers.
+So the prompt is also very simple and straightforward.
+
+```
+Q: Can you write me a rust function that compute the integer average of two integers?
+```
+
+Giving the above prompt, ChatGPT answered us with the following function
+```rust
+fn integer_average(a: i32, b: i32) -> i32 {
+    (a + b) / 2
+}
+```
+, the explanation of the function
+>In this function, we define a function called integer_average that takes two integer arguments a and b.
+The function then adds the two arguments together and divides the result by 2 to compute the integer average.
+
+, and even a drive function to call the average function.
+```rust
+fn main() {
+    let a = 4;
+    let b = 8;
+    let result = integer_average(a, b);
+    println!("The average of {} and {} is {}", a, b, result);
+}
+```
+
+#### Example 2.a
```

Comment:
Naming the examples in a different way (e.g., the "average example" and the "lowest unset bit example" if you want to focus on their purpose, or the "program writing example" and "optimization example" if you want to focus on the task) may help readers in remembering these examples later. But I'm not pushing too hard on this; I just think "2.a" or "1.b" is easy to forget.